{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "656a2484-811f-4fa4-af06-02e6ba1430e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import msprime\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "import time\n",
    "from IPython.display import SVG, display\n",
    "\n",
    "class MyVcfSim:\n",
    "    \n",
    "    def __init__(self, site_size, ploidy, pop_num, mutationrate, percentmissing, randoseed, outputfile, samp_num, samp_file, folder):\n",
    "        \n",
    "        self.site_size = site_size\n",
    "        self.ploidy = ploidy\n",
    "        self.pop_num = pop_num\n",
    "        self.mutationrate = mutationrate\n",
    "        self.percentmissing = percentmissing\n",
    "        self.randoseed = randoseed\n",
    "        self.outputfile = outputfile\n",
    "        self.samp_num = samp_num\n",
    "        self.samp_file = samp_file\n",
    "        self.folder = folder\n",
    "    \n",
    "    def make_site_mask(self):\n",
    "\n",
    "        temp_percent = self.percentmissing/100 #takes percent of user input and converts it to actual percentage\n",
    "        temp_var = self.site_size*temp_percent  #finds amount of VCF's that need to be taken out\n",
    "        counter = range(int(temp_var)) #Makes a counter to traverse VCF's\n",
    "        temp_array = np.zeros(self.site_size, dtype = int) #Makes array of 0's \n",
    "        \n",
    "        mylist = np.arange(0, self.site_size,1) #Generates an array of 1-size_size\n",
    "        rand_array = np.random.choice(mylist, self.site_size, replace=False) #fills a new array with random non repeating values from last step\n",
    "        \n",
    "\n",
    "        print('Total Amount:', temp_array.size)  #Shows total amount of sites\n",
    "\n",
    "        print('----------------')\n",
    "\n",
    "        for x in counter:\n",
    "            temp_array[rand_array[x]] = 1 #Changes value to 1 which deletes the value\n",
    "\n",
    "        print('Amount Missing:', np.sum(temp_array))\n",
    "\n",
    "        print('----------------')\n",
    "        \n",
    "        rand_array = np.empty(self.site_size, dtype = int)\n",
    "                \n",
    "        return temp_array\n",
    "\n",
    "    def make_missing_vcf(self, ts):\n",
    "\n",
    "        site_mask = self.make_site_mask() #Makes sites from the number of sites in ts\n",
    "\n",
    "        with open(self.outputfile, \"w\") as f:\n",
    "            ts.write_vcf(f, site_mask=site_mask)\n",
    "\n",
    "    def make_population_file(self):\n",
    "\n",
    "        file = open(self.samp_file, \"a\") #Makes a text file full of population data as a parameter for pixy\n",
    "\n",
    "        for x in range(0, self.samp_num):\n",
    "            file.write(\"tsk_\"+ str(x) + \"\\t1\\n\")\n",
    "        file.close()\n",
    "\n",
    "    def simulate_vcfs(self):\n",
    "        \n",
    "        ts = msprime.sim_ancestry(samples=[msprime.SampleSet(self.samp_num, ploidy=self.ploidy)], population_size = self.pop_num, random_seed=self.randoseed, sequence_length = self.site_size)\n",
    "\n",
    "        #sequence length and site size have to be set to SAME number\n",
    "\n",
    "        #sets number of sites\n",
    "\n",
    "        rlg = np.random.default_rng(self.randoseed) #random letter generator \n",
    "\n",
    "        rchar = rlg.integers(low=0, high=4) #randomly choses number from 0 to 3 which represents genome\n",
    "\n",
    "        difference_counter = range(self.site_size) \n",
    "\n",
    "        tables = ts.dump_tables() \n",
    "\n",
    "        for x in difference_counter: #Randomly assings reference genome for each site\n",
    "            \n",
    "            tables.sites.add_row(x, \"A\") \n",
    "\n",
    "        ts = tables.tree_sequence()\n",
    "\n",
    "        ts = msprime.sim_mutations(ts, rate=self.mutationrate, random_seed=self.randoseed) #Mutates the sites at random\n",
    "\n",
    "        self.make_missing_vcf(ts)\n",
    "\n",
    "        #os.system('cat my.vcf | grep \"^#\" > vcf_new.vcf')\n",
    "        #os.system('cat my.vcf | grep -v \"^#\" | awk -v s=1 \\'{$2=$2+s; $3=$3+s; print}\\' >> vcf_new.vcf')\n",
    "        #os.system('rm my.vcf')\n",
    "\n",
    "    def simulate_pix_once(self):\n",
    "        #Function meant to test pixy once, keeping all the files\n",
    "        defaultdirectory = os.getcwd() \n",
    "\n",
    "        newfolder = self.folder\n",
    "\n",
    "        os.mkdir(newfolder) \n",
    "        os.chdir(newfolder) #Creates new directory to store data files\n",
    "\n",
    "        self.make_population_file()\n",
    "        self.simulate_vcfs() \n",
    "\n",
    "        command1 = 'bgzip ' + self.outputfile\n",
    "        command2 = 'tabix -f -p vcf -0 ' + self.outputfile + '.gz'\n",
    "        command3 = 'pixy --stats pi --vcf ' + self.outputfile + '.gz --populations '+ self.samp_file + ' --window_size 10000 --output_prefix PixyTest' + str(self.randoseed) \n",
    "\n",
    "        #Commands used to make text files in the proper format to be inputed into pixy\n",
    "        \n",
    "        os.system(command1)\n",
    "        os.system(command2)\n",
    "        os.system(command3)\n",
    "        \n",
    "        os.chdir(defaultdirectory) \n",
    "            \n",
    "    def mass_simulate_pix(self):\n",
    "        #Function meant to test pixy multiple times. All files except for the main pixy file are deleted. \n",
    "        self.make_population_file()\n",
    "        self.simulate_vcfs()\n",
    "\n",
    "        command1 = 'bgzip ' + self.outputfile\n",
    "        command2 = 'tabix -f -p vcf -0 ' + self.outputfile + '.gz'\n",
    "        command3 = 'pixy --stats pi --vcf ' + self.outputfile + '.gz --populations '+ self.samp_file + ' --window_size 10000 --output_prefix PixyTest' + str(self.randoseed) \n",
    "        command4 = self.outputfile + '.gz'\n",
    "        command5 = self.outputfile + '.gz.tbi'\n",
    "        command6 = self.samp_file\n",
    "        \n",
    "        #Commands used to make text files in the proper format to be inputed into pixy. Deletes excess files aswell. \n",
    "        \n",
    "        os.system(command1)\n",
    "        os.system(command2)\n",
    "        os.system(command3)\n",
    "        os.remove(command4)\n",
    "        os.remove(command5)\n",
    "        os.remove(command6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e4f38ef5-dbdb-42ee-aed3-8b8182cb86cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function meant to simulate pixy on a mass scale, outputting all data to a .csv data sheet\n",
    "def mass_simulator(sitesize, ploidy, popnum, mutationrate,\n",
    "                   percentmissing, seed, outputfile, sampnum, sampfile, foldername, amountofruns = 1):\n",
    "    \n",
    "    columns = ('ploidy', 'mutation_rate', 'percent_missing', 'seed')\n",
    "    df = pd.DataFrame(index = columns).T #Creates a pandas data frame with the columns above.\n",
    "    \n",
    "    defaultdirectory = os.getcwd() \n",
    "    os.mkdir(foldername) \n",
    "    os.chdir(foldername) \n",
    "\n",
    "    for x in range(amountofruns):\n",
    "        \n",
    "        foldername = 'folder' + str(seed)\n",
    "        sim = MyVcfSim(sitesize, ploidy, popnum, mutationrate, percentmissing, seed, \n",
    "                       outputfile, sampnum, sampfile, foldername)\n",
    "        \n",
    "        #Runs simulation with the data from above.\n",
    "        \n",
    "        sim.mass_simulate_pix()\n",
    "                \n",
    "        pixdata = pd.read_csv('PixyTest' + str(seed) + '_pi.txt', sep=\"\\t\")\n",
    "        \n",
    "        #Uses pandas to read data given from pixy into another data frame\n",
    "        \n",
    "        data = {\n",
    "        0: {'ploidy': sim.ploidy, 'mutation_rate': sim.mutationrate, 'percent_missing': sim.percentmissing, \n",
    "                          'seed': seed},\n",
    "        }\n",
    "        \n",
    "        df2 = pd.DataFrame(data=data, index = columns).T\n",
    "        \n",
    "        df2 = pd.concat([df2,pixdata], axis = 1) \n",
    "        \n",
    "        df = pd.concat([df, df2], axis = 0)\n",
    "        \n",
    "        #Combines the two data frames together side by side so all the pixy data matches up to the correct test run in a big data frame\n",
    "        \n",
    "        seed+=1\n",
    "        \n",
    "    os.chdir(defaultdirectory) \n",
    "    df.to_csv('data.csv')\n",
    "    \n",
    "    #Outputs all the data from the main frame into its own csv file\n",
    "    \n",
    "    \n",
    "#Quick function to test if VCF files are getting outputed correctly\n",
    "def vcf_simulator(amountofruns = 1, seed = 1000, foldername = 'PixyFolder'):\n",
    "    \n",
    "    for x in range(amountofruns):\n",
    "        outputfile = 'myvcf' + str(seed)\n",
    "        foldername = 'folder' + str(seed)\n",
    "        sim = MyVcfSim(10, 1, 1700000, 0.0000001, 20, seed, outputfile, 20, 'population.txt', 'vcff')\n",
    "        sim.simulate_vcfs()\n",
    "        seed+=1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2686f419-8305-44f5-a7b1-e1cb50ab28c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Amount: 10000\n",
      "----------------\n",
      "Amount Missing: 2000\n",
      "----------------\n",
      "[pixy] pixy 1.2.7.beta1\n",
      "[pixy] See documentation at https://pixy.readthedocs.io/en/latest/\n",
      "\n",
      "[pixy] Validating VCF and input parameters...\n",
      "[pixy] Checking write access...OK\n",
      "[pixy] Checking CPU configuration...OK\n",
      "[pixy] Checking for invariant sites...OK\n",
      "[pixy] Checking chromosome data...OK\n",
      "[pixy] Checking intervals/sites...OK\n",
      "[pixy] Checking sample data...OK\n",
      "[pixy] All initial checks past!\n",
      "\n",
      "[pixy] Preparing for calculation of summary statistics: pi\n",
      "[pixy] Data set contains 1 population(s), 1 chromosome(s), and 20 sample(s)\n",
      "[pixy] Window size: 10000 bp\n",
      "\n",
      "[pixy] Started calculations at 19:56:35 on 2022-10-28\n",
      "[pixy] Using 1 out of 4 available CPU cores\n",
      "\n",
      "[pixy] Processing chromosome/contig 1...\n",
      "[pixy] Calculating statistics for region 1:1-9999...\n",
      "\n",
      "[pixy] All calculations complete at 19:56:35 on 2022-10-28\n",
      "[pixy] Time elapsed: 00:00:00\n",
      "[pixy] Output files written to: /Users/paimongoulart/MyBetterProject/MyProject/TestNotebooks/PixyFolder/\n",
      "\n",
      "[pixy] If you use pixy in your research, please cite the following paper:\n",
      "[pixy] Korunes, KL and K Samuk. pixy: Unbiased estimation of nucleotide diversity and divergence in the presence of missing data. Mol Ecol Resour. 2021 Jan 16. doi: 10.1111/1755-0998.13326.\n",
      "Total Amount: 10000\n",
      "----------------\n",
      "Amount Missing: 2000\n",
      "----------------\n",
      "[pixy] pixy 1.2.7.beta1\n",
      "[pixy] See documentation at https://pixy.readthedocs.io/en/latest/\n",
      "\n",
      "[pixy] Validating VCF and input parameters...\n",
      "[pixy] Checking write access...OK\n",
      "[pixy] Checking CPU configuration...OK\n",
      "[pixy] Checking for invariant sites...OK\n",
      "[pixy] Checking chromosome data...OK\n",
      "[pixy] Checking intervals/sites...OK\n",
      "[pixy] Checking sample data...OK\n",
      "[pixy] All initial checks past!\n",
      "\n",
      "[pixy] Preparing for calculation of summary statistics: pi\n",
      "[pixy] Data set contains 1 population(s), 1 chromosome(s), and 20 sample(s)\n",
      "[pixy] Window size: 10000 bp\n",
      "\n",
      "[pixy] Started calculations at 19:56:38 on 2022-10-28\n",
      "[pixy] Using 1 out of 4 available CPU cores\n",
      "\n",
      "[pixy] Processing chromosome/contig 1...\n",
      "[pixy] Calculating statistics for region 1:1-9999...\n",
      "\n",
      "[pixy] All calculations complete at 19:56:38 on 2022-10-28\n",
      "[pixy] Time elapsed: 00:00:00\n",
      "[pixy] Output files written to: /Users/paimongoulart/MyBetterProject/MyProject/TestNotebooks/PixyFolder/\n",
      "\n",
      "[pixy] If you use pixy in your research, please cite the following paper:\n",
      "[pixy] Korunes, KL and K Samuk. pixy: Unbiased estimation of nucleotide diversity and divergence in the presence of missing data. Mol Ecol Resour. 2021 Jan 16. doi: 10.1111/1755-0998.13326.\n",
      "Total Amount: 10000\n",
      "----------------\n",
      "Amount Missing: 2000\n",
      "----------------\n",
      "[pixy] pixy 1.2.7.beta1\n",
      "[pixy] See documentation at https://pixy.readthedocs.io/en/latest/\n",
      "\n",
      "[pixy] Validating VCF and input parameters...\n",
      "[pixy] Checking write access...OK\n",
      "[pixy] Checking CPU configuration...OK\n",
      "[pixy] Checking for invariant sites...OK\n",
      "[pixy] Checking chromosome data...OK\n",
      "[pixy] Checking intervals/sites...OK\n",
      "[pixy] Checking sample data...OK\n",
      "[pixy] All initial checks past!\n",
      "\n",
      "[pixy] Preparing for calculation of summary statistics: pi\n",
      "[pixy] Data set contains 1 population(s), 1 chromosome(s), and 20 sample(s)\n",
      "[pixy] Window size: 10000 bp\n",
      "\n",
      "[pixy] Started calculations at 19:56:41 on 2022-10-28\n",
      "[pixy] Using 1 out of 4 available CPU cores\n",
      "\n",
      "[pixy] Processing chromosome/contig 1...\n",
      "[pixy] Calculating statistics for region 1:1-9999...\n",
      "\n",
      "[pixy] All calculations complete at 19:56:41 on 2022-10-28\n",
      "[pixy] Time elapsed: 00:00:00\n",
      "[pixy] Output files written to: /Users/paimongoulart/MyBetterProject/MyProject/TestNotebooks/PixyFolder/\n",
      "\n",
      "[pixy] If you use pixy in your research, please cite the following paper:\n",
      "[pixy] Korunes, KL and K Samuk. pixy: Unbiased estimation of nucleotide diversity and divergence in the presence of missing data. Mol Ecol Resour. 2021 Jan 16. doi: 10.1111/1755-0998.13326.\n",
      "Total Amount: 10000\n",
      "----------------\n",
      "Amount Missing: 2000\n",
      "----------------\n",
      "[pixy] pixy 1.2.7.beta1\n",
      "[pixy] See documentation at https://pixy.readthedocs.io/en/latest/\n",
      "\n",
      "[pixy] Validating VCF and input parameters...\n",
      "[pixy] Checking write access...OK\n",
      "[pixy] Checking CPU configuration...OK\n",
      "[pixy] Checking for invariant sites...OK\n",
      "[pixy] Checking chromosome data...OK\n",
      "[pixy] Checking intervals/sites...OK\n",
      "[pixy] Checking sample data...OK\n",
      "[pixy] All initial checks past!\n",
      "\n",
      "[pixy] Preparing for calculation of summary statistics: pi\n",
      "[pixy] Data set contains 1 population(s), 1 chromosome(s), and 20 sample(s)\n",
      "[pixy] Window size: 10000 bp\n",
      "\n",
      "[pixy] Started calculations at 19:56:44 on 2022-10-28\n",
      "[pixy] Using 1 out of 4 available CPU cores\n",
      "\n",
      "[pixy] Processing chromosome/contig 1...\n",
      "[pixy] Calculating statistics for region 1:1-9999...\n",
      "\n",
      "[pixy] All calculations complete at 19:56:44 on 2022-10-28\n",
      "[pixy] Time elapsed: 00:00:00\n",
      "[pixy] Output files written to: /Users/paimongoulart/MyBetterProject/MyProject/TestNotebooks/PixyFolder/\n",
      "\n",
      "[pixy] If you use pixy in your research, please cite the following paper:\n",
      "[pixy] Korunes, KL and K Samuk. pixy: Unbiased estimation of nucleotide diversity and divergence in the presence of missing data. Mol Ecol Resour. 2021 Jan 16. doi: 10.1111/1755-0998.13326.\n",
      "Total Amount: 10000\n",
      "----------------\n",
      "Amount Missing: 2000\n",
      "----------------\n",
      "[pixy] pixy 1.2.7.beta1\n",
      "[pixy] See documentation at https://pixy.readthedocs.io/en/latest/\n",
      "\n",
      "[pixy] Validating VCF and input parameters...\n",
      "[pixy] Checking write access...OK\n",
      "[pixy] Checking CPU configuration...OK\n",
      "[pixy] Checking for invariant sites...OK\n",
      "[pixy] Checking chromosome data...OK\n",
      "[pixy] Checking intervals/sites...OK\n",
      "[pixy] Checking sample data...OK\n",
      "[pixy] All initial checks past!\n",
      "\n",
      "[pixy] Preparing for calculation of summary statistics: pi\n",
      "[pixy] Data set contains 1 population(s), 1 chromosome(s), and 20 sample(s)\n",
      "[pixy] Window size: 10000 bp\n",
      "\n",
      "[pixy] Started calculations at 19:56:47 on 2022-10-28\n",
      "[pixy] Using 1 out of 4 available CPU cores\n",
      "\n",
      "[pixy] Processing chromosome/contig 1...\n",
      "[pixy] Calculating statistics for region 1:1-9998...\n",
      "\n",
      "[pixy] All calculations complete at 19:56:47 on 2022-10-28\n",
      "[pixy] Time elapsed: 00:00:00\n",
      "[pixy] Output files written to: /Users/paimongoulart/MyBetterProject/MyProject/TestNotebooks/PixyFolder/\n",
      "\n",
      "[pixy] If you use pixy in your research, please cite the following paper:\n",
      "[pixy] Korunes, KL and K Samuk. pixy: Unbiased estimation of nucleotide diversity and divergence in the presence of missing data. Mol Ecol Resour. 2021 Jan 16. doi: 10.1111/1755-0998.13326.\n",
      "Total Amount: 10000\n",
      "----------------\n",
      "Amount Missing: 2000\n",
      "----------------\n",
      "[pixy] pixy 1.2.7.beta1\n",
      "[pixy] See documentation at https://pixy.readthedocs.io/en/latest/\n",
      "\n",
      "[pixy] Validating VCF and input parameters...\n",
      "[pixy] Checking write access...OK\n",
      "[pixy] Checking CPU configuration...OK\n",
      "[pixy] Checking for invariant sites...OK\n",
      "[pixy] Checking chromosome data...OK\n",
      "[pixy] Checking intervals/sites...OK\n",
      "[pixy] Checking sample data...OK\n",
      "[pixy] All initial checks past!\n",
      "\n",
      "[pixy] Preparing for calculation of summary statistics: pi\n",
      "[pixy] Data set contains 1 population(s), 1 chromosome(s), and 20 sample(s)\n",
      "[pixy] Window size: 10000 bp\n",
      "\n",
      "[pixy] Started calculations at 19:56:50 on 2022-10-28\n",
      "[pixy] Using 1 out of 4 available CPU cores\n",
      "\n",
      "[pixy] Processing chromosome/contig 1...\n",
      "[pixy] Calculating statistics for region 1:1-9999...\n",
      "\n",
      "[pixy] All calculations complete at 19:56:50 on 2022-10-28\n",
      "[pixy] Time elapsed: 00:00:00\n",
      "[pixy] Output files written to: /Users/paimongoulart/MyBetterProject/MyProject/TestNotebooks/PixyFolder/\n",
      "\n",
      "[pixy] If you use pixy in your research, please cite the following paper:\n",
      "[pixy] Korunes, KL and K Samuk. pixy: Unbiased estimation of nucleotide diversity and divergence in the presence of missing data. Mol Ecol Resour. 2021 Jan 16. doi: 10.1111/1755-0998.13326.\n",
      "Total Amount: 10000\n",
      "----------------\n",
      "Amount Missing: 2000\n",
      "----------------\n",
      "[pixy] pixy 1.2.7.beta1\n",
      "[pixy] See documentation at https://pixy.readthedocs.io/en/latest/\n",
      "\n",
      "[pixy] Validating VCF and input parameters...\n",
      "[pixy] Checking write access...OK\n",
      "[pixy] Checking CPU configuration...OK\n",
      "[pixy] Checking for invariant sites...OK\n",
      "[pixy] Checking chromosome data...OK\n",
      "[pixy] Checking intervals/sites...OK\n",
      "[pixy] Checking sample data...OK\n",
      "[pixy] All initial checks past!\n",
      "\n",
      "[pixy] Preparing for calculation of summary statistics: pi\n",
      "[pixy] Data set contains 1 population(s), 1 chromosome(s), and 20 sample(s)\n",
      "[pixy] Window size: 10000 bp\n",
      "\n",
      "[pixy] Started calculations at 19:56:53 on 2022-10-28\n",
      "[pixy] Using 1 out of 4 available CPU cores\n",
      "\n",
      "[pixy] Processing chromosome/contig 1...\n",
      "[pixy] Calculating statistics for region 1:1-9999...\n",
      "\n",
      "[pixy] All calculations complete at 19:56:53 on 2022-10-28\n",
      "[pixy] Time elapsed: 00:00:00\n",
      "[pixy] Output files written to: /Users/paimongoulart/MyBetterProject/MyProject/TestNotebooks/PixyFolder/\n",
      "\n",
      "[pixy] If you use pixy in your research, please cite the following paper:\n",
      "[pixy] Korunes, KL and K Samuk. pixy: Unbiased estimation of nucleotide diversity and divergence in the presence of missing data. Mol Ecol Resour. 2021 Jan 16. doi: 10.1111/1755-0998.13326.\n",
      "Total Amount: 10000\n",
      "----------------\n",
      "Amount Missing: 2000\n",
      "----------------\n",
      "[pixy] pixy 1.2.7.beta1\n",
      "[pixy] See documentation at https://pixy.readthedocs.io/en/latest/\n",
      "\n",
      "[pixy] Validating VCF and input parameters...\n",
      "[pixy] Checking write access...OK\n",
      "[pixy] Checking CPU configuration...OK\n",
      "[pixy] Checking for invariant sites...OK\n",
      "[pixy] Checking chromosome data...OK\n",
      "[pixy] Checking intervals/sites...OK\n",
      "[pixy] Checking sample data...OK\n",
      "[pixy] All initial checks past!\n",
      "\n",
      "[pixy] Preparing for calculation of summary statistics: pi\n",
      "[pixy] Data set contains 1 population(s), 1 chromosome(s), and 20 sample(s)\n",
      "[pixy] Window size: 10000 bp\n",
      "\n",
      "[pixy] Started calculations at 19:56:56 on 2022-10-28\n",
      "[pixy] Using 1 out of 4 available CPU cores\n",
      "\n",
      "[pixy] Processing chromosome/contig 1...\n",
      "[pixy] Calculating statistics for region 1:1-9998...\n",
      "\n",
      "[pixy] All calculations complete at 19:56:56 on 2022-10-28\n",
      "[pixy] Time elapsed: 00:00:00\n",
      "[pixy] Output files written to: /Users/paimongoulart/MyBetterProject/MyProject/TestNotebooks/PixyFolder/\n",
      "\n",
      "[pixy] If you use pixy in your research, please cite the following paper:\n",
      "[pixy] Korunes, KL and K Samuk. pixy: Unbiased estimation of nucleotide diversity and divergence in the presence of missing data. Mol Ecol Resour. 2021 Jan 16. doi: 10.1111/1755-0998.13326.\n",
      "Total Amount: 10000\n",
      "----------------\n",
      "Amount Missing: 2000\n",
      "----------------\n",
      "[pixy] pixy 1.2.7.beta1\n",
      "[pixy] See documentation at https://pixy.readthedocs.io/en/latest/\n",
      "\n",
      "[pixy] Validating VCF and input parameters...\n",
      "[pixy] Checking write access...OK\n",
      "[pixy] Checking CPU configuration...OK\n",
      "[pixy] Checking for invariant sites...OK\n",
      "[pixy] Checking chromosome data...OK\n",
      "[pixy] Checking intervals/sites...OK\n",
      "[pixy] Checking sample data...OK\n",
      "[pixy] All initial checks past!\n",
      "\n",
      "[pixy] Preparing for calculation of summary statistics: pi\n",
      "[pixy] Data set contains 1 population(s), 1 chromosome(s), and 20 sample(s)\n",
      "[pixy] Window size: 10000 bp\n",
      "\n",
      "[pixy] Started calculations at 19:56:59 on 2022-10-28\n",
      "[pixy] Using 1 out of 4 available CPU cores\n",
      "\n",
      "[pixy] Processing chromosome/contig 1...\n",
      "[pixy] Calculating statistics for region 1:1-9999...\n",
      "\n",
      "[pixy] All calculations complete at 19:56:59 on 2022-10-28\n",
      "[pixy] Time elapsed: 00:00:00\n",
      "[pixy] Output files written to: /Users/paimongoulart/MyBetterProject/MyProject/TestNotebooks/PixyFolder/\n",
      "\n",
      "[pixy] If you use pixy in your research, please cite the following paper:\n",
      "[pixy] Korunes, KL and K Samuk. pixy: Unbiased estimation of nucleotide diversity and divergence in the presence of missing data. Mol Ecol Resour. 2021 Jan 16. doi: 10.1111/1755-0998.13326.\n",
      "Total Amount: 10000\n",
      "----------------\n",
      "Amount Missing: 2000\n",
      "----------------\n",
      "[pixy] pixy 1.2.7.beta1\n",
      "[pixy] See documentation at https://pixy.readthedocs.io/en/latest/\n",
      "\n",
      "[pixy] Validating VCF and input parameters...\n",
      "[pixy] Checking write access...OK\n",
      "[pixy] Checking CPU configuration...OK\n",
      "[pixy] Checking for invariant sites...OK\n",
      "[pixy] Checking chromosome data...OK\n",
      "[pixy] Checking intervals/sites...OK\n",
      "[pixy] Checking sample data...OK\n",
      "[pixy] All initial checks past!\n",
      "\n",
      "[pixy] Preparing for calculation of summary statistics: pi\n",
      "[pixy] Data set contains 1 population(s), 1 chromosome(s), and 20 sample(s)\n",
      "[pixy] Window size: 10000 bp\n",
      "\n",
      "[pixy] Started calculations at 19:57:01 on 2022-10-28\n",
      "[pixy] Using 1 out of 4 available CPU cores\n",
      "\n",
      "[pixy] Processing chromosome/contig 1...\n",
      "[pixy] Calculating statistics for region 1:1-9999...\n",
      "\n",
      "[pixy] All calculations complete at 19:57:02 on 2022-10-28\n",
      "[pixy] Time elapsed: 00:00:00\n",
      "[pixy] Output files written to: /Users/paimongoulart/MyBetterProject/MyProject/TestNotebooks/PixyFolder/\n",
      "\n",
      "[pixy] If you use pixy in your research, please cite the following paper:\n",
      "[pixy] Korunes, KL and K Samuk. pixy: Unbiased estimation of nucleotide diversity and divergence in the presence of missing data. Mol Ecol Resour. 2021 Jan 16. doi: 10.1111/1755-0998.13326.\n"
     ]
    }
   ],
   "source": [
    "mass_simulator(10000, 2, 1700000, 0.0000001, 20, 1000, 'my.vcf', 20, 'population.txt', 'PixyFolder', 10)\n",
    "#Mass simulator function to run everything above.\n",
    "#Parameters in order: Site Size, Ploidy, Population, Mutation Rate, Percent Of Missing data to be taken out,-\n",
    "#Starting seed, Output VCF file, Sample Size, Sample file, Output Folder, Amount of times to run everything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f612febe-6d9d-45a6-ab83-0d40d8ac8707",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
